### DISCUSSION TITLE
--word_timestamps True lead to Failed to launch Triton kernels warning

### DISCUSSION THREAD

**Discussion by zx3777** [openai/whisper#1283] [General]
i use 3060ti 8g gpu ，win10. cuda 11.7

when i use --word_timestamps True ，it will show this twice ：

C:\Users\xx\AppData\Local\Programs\Python\Python310\lib\site-packages\whisper\timing.py:42: UserWarning: Failed to launch Triton kernels, likely due to missing CUDA toolkit; falling back to a slower median kernel implementation...
  warnings.warn(
C:\Users\xx\AppData\Local\Programs\Python\Python310\lib\site-packages\whisper\timing.py:146: UserWarning: Failed to launch Triton kernels, likely due to missing CUDA toolkit; falling back to a slower DTW implementation...

But the task can be executed and the correct srt can be obtained. How to fix this warning?

**Comment by sm1045** [comment]
Same problem here

**Comment by himberjack** [comment]
Same here

**Comment by jongwook** [comment]
This is because [triton does not provide Windows binary](https://pypi.org/project/triton/#files), and I'm not sure if it's planned to do so. I should probably not print the warning on platforms other than x86_64 linux..

**Reply by himberjack** [reply]
Happens to me with cuda linux docker

**Reply by jongwook** [reply]
Triton requires CUDA toolkit installed, which you can confirm by checking if `nvcc` is available in the CLI. If not, you can get them by switching to [a "devel" version of the Docker image](https://hub.docker.com/r/nvidia/cuda/tags?page=1&name=devel), but it'll take a couple of gigabytes more.

**Reply by sm1045** [reply]
It happens to me in linux tegra with CUDA toolkit installed. I am using Jetson Orin Agx to run whisper in it.

**This is the output of nvcc.**
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2022 NVIDIA Corporation
Built on Sun_Oct_23_22:16:07_PDT_2022
Cuda compilation tools, release 11.4, V11.4.315
Build cuda_11.4.r11.4/compiler.31964100_0

**More info about CUDA**
2.0.0a0+8aa34602.nv23.03
11.4
8600
Orin
_CudaDeviceProperties(name='Orin', major=8, minor=7, total_memory=30587MB, multi_processor_count=8)
11.4
torch.Size([1, 3, 222, 222])

**Pytorch info**
Collecting environment information...
PyTorch version: 2.0.0a0+8aa34602.nv23.03
Is debug build: False
CUDA used to build PyTorch: 11.4
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.6 LTS (aarch64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.31

Python version: 3.8.10 (default, Mar 13 2023, 10:26:41)  [GCC 9.4.0] (64-bit runtime)
Python platform: Linux-5.10.104-tegra-aarch64-with-glibc2.29
Is CUDA available: True
CUDA runtime version: 11.4.315
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect

**Reply by jongwook** [reply]
Thanks for more information, and yes unfortunately Triton does not provide wheels for aarch64 linux either. Seems like it's technically possible to build the package [which may involve rebuilding LLVM for aarch64](https://github.com/openai/triton/issues/1495). If the numba version works, it shouldn't be too slow while avoiding the hassle of building Triton.

**Reply by sm1045** [reply]
Thx for this information @jongwook! I will research a little more about this.

**Reply by RichoDS** [reply]
@jongwook Can i do simultaneous 2 language transcription on openai whisper?

**Reply by ryanheise** [reply]
@RichoDS this discussion is about the Triton warning, but you can try searching the discussions for your topic to find more relevant discussions. I did a search and found these: #807 #49 #849 #345 #1318 . You can pick the discussion that has attracted the most useful comments and continue that discussion, since (as you can see) it might not be convenient to continue the discussion in 5 different places.

**Reply by JosemyL95** [reply]
@zx3777 @jongwook @ryanheise 
Hi guys! Do u know how can we disable the triton warnings? I'm actually using language transcriptions (with word timestamps) in Windows. I've already checked official docs and didn't saw any tip to disable the warning.
Thanks and greetings!

**Reply by Honghe** [reply]
> This is because [triton does not provide Windows binary](https://pypi.org/project/triton/#files), and I'm not sure if it's planned to do so. I should probably not print the warning on platforms other than x86_64 linux..

There is a Windows port of triton [Windows 11 cannot install Triton · Issue #4595 · triton-lang/triton](https://github.com/triton-lang/triton/issues/4595).
Even if triton-windows is installed, There is no speed up on a short audi, e.g. 2 minutes, the elapsed time is 20 seconds on rtx 4050 laptop.

**Comment by aheba** [comment]
I use ubuntu, CUDA 11.7 and have the same issue +1

**Reply by taras-sereda** [reply]
same here

**Reply by zhongxinlei** [reply]
same here, I also using Ubuntu.

**Comment by okineadev** [comment]
### I seem to have found a fix for the problem, here are the steps I followed:

1. Downloaded and installed the **CUDA** driver (**12.1**) for **WSL** from this page: https://developer.nvidia.com/cuda-12-1-0-download-archive
  > [!NOTE]
  > I don't know if version **12.1** is really needed, but I installed it because pytorch comes with **CUDA 12.1** support 🤷‍♂️
  >
  > UPD: Still works with newer versions, including the latest (12.4)
  >
  > ![image](https://github.com/openai/whisper/assets/81070564/788f493d-c4f5-4d2a-989d-9dbe1cac2d1b)

2. Added **CUDA** location to **$PATH**:

> **~/.bashrc**

```bash
export PATH="$PATH:/usr/local/cuda/bin"
```

3. Reboot

```bash
sudo reboot
```

---

When I installed all this - the error did not go away, **BUT** - I found a solution!
Here is a link to the solution - https://stackoverflow.com/a/76380867/21165921

4. Installed `libnvidia-extra-530` package

```bash
sudo apt update
sudo apt install libnvidia-extra-530
```

---

After that the error with **Triton** disappeared :)

**Reply by doggy8088** [reply]
Hi, Do you install CUDA driver (12.1) in the WSL environment? In WSL, there is no way to reboot. I don't understand how you do that?

**Reply by okineadev** [reply]
> Hi, Do you install CUDA driver (12.1) in the WSL environment? In WSL, there is no way to reboot. I don't understand how you do that?

```bash
wsl --shutdown
wsl --shutdown
```

**Reply by doggy8088** [reply]
Should I install CUDA toolkit on "Windows"? Or just install in the WSL only?

**Reply by doggy8088** [reply]
The CUDA download page seems changed. I don't know how to choose. Should I choose "Windows"? In your screenshot, you seems choose the "Windows" OS. I don't understand.

![image](https://github.com/user-attachments/assets/21f934d9-42ea-4e7b-8b5b-6a8d222bb8c4)

**Reply by okineadev** [reply]
> The CUDA download page seems changed. I don't know how to choose. Should I choose "Windows"? In your screenshot, you seems choose the "Windows" OS. I don't understand.
> 
> ![image](https://private-user-images.githubusercontent.com/88981/374650143-21f934d9-42ea-4e7b-8b5b-6a8d222bb8c4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg0NDcxODIsIm5iZiI6MTcyODQ0Njg4MiwicGF0aCI6Ii84ODk4MS8zNzQ2NTAxNDMtMjFmOTM0ZDktNDJlYS00ZTdiLThiNWItNmE4ZDIyMmJiOGM0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDA5VDA0MDgwMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk3MTYzODQwMTQ0YTAxZGJjYzY4ZTcxMjkxOTliMjVjMzZmZDEwYjYyMmUyNGFhYTZiN2ZkZTljNzdhODU4NGQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Zej3OQnOiNdsvN_Fjjh5UqPldwyXrOWB_lUrhALxkzw)

https://developer.nvidia.com/cuda-12-1-0-download-archive?target_os=Linux&target_arch=x86_64&Distribution=WSL-Ubuntu

**Reply by doggy8088** [reply]
In [this page](https://docs.nvidia.com/cuda/wsl-user-guide/index.html#cuda-support-for-wsl-2), NVIDIA said **users must not install any NVIDIA GPU Linux driver within WSL 2**.

![image](https://github.com/user-attachments/assets/1c9ada67-1656-4e87-9d98-ad61c9c8ca33)

**Reply by ramondsq** [reply]
> In [this page](https://docs.nvidia.com/cuda/wsl-user-guide/index.html#cuda-support-for-wsl-2), NVIDIA said **users must not install any NVIDIA GPU Linux driver within WSL 2**.
> 
> ![image](https://private-user-images.githubusercontent.com/88981/375219745-1c9ada67-1656-4e87-9d98-ad61c9c8ca33.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDI1NzIzNjYsIm5iZiI6MTc0MjU3MjA2NiwicGF0aCI6Ii84ODk4MS8zNzUyMTk3NDUtMWM5YWRhNjctMTY1Ni00ZTg3LTlkOTgtYWQ2MWM5YzhjYTMzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzIxVDE1NDc0NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWUwZmI3MmQyYjVjODYwMjdlMzM3ZTczYzI5MDVmYzg1YmIwNTRhZWJmY2JjOGMwNTg2YzYwNTAyMTQzMTIyMzUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.GkxexZYKhcgSDhi9JojzIhG52lALkDrL56gbil4hLCY)

GPU driver is not the same thing as cuda toolkit
